File: app/src/main/java/com/stedi/randomimagegenerator/app/view/fragments/ChooseEffectFragment.java
package com.stedi.randomimagegenerator.app.view.fragments;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.stedi.randomimagegenerator.app.R;
import com.stedi.randomimagegenerator.app.model.data.GeneratorType;
import com.stedi.randomimagegenerator.app.other.logger.Logger;
import com.stedi.randomimagegenerator.app.presenter.interfaces.ChooseEffectPresenter;
import com.stedi.randomimagegenerator.app.view.activity.GenerationStepsActivity;
import com.stedi.randomimagegenerator.app.view.adapters.GeneratorTypeAdapter;
import com.stedi.randomimagegenerator.app.view.components.GeneratorTypeImageLoader;
import com.stedi.randomimagegenerator.app.view.fragments.base.StepFragment;

import javax.inject.Inject;

import butterknife.BindView;

public class ChooseEffectFragment extends StepFragment implements
        GeneratorTypeAdapter.ClickListener,
        ChooseEffectPresenter.UIImpl {

    @Inject ChooseEffectPresenter presenter;
    @Inject Logger logger;
    @Inject GeneratorTypeImageLoader adapterImageLoader;

    @BindView(R.id.choose_effect_fragment_recycler_view) RecyclerView recyclerView;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((GenerationStepsActivity) getActivity()).getGenerationComponent().inject(this);
        presenter.onAttach(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        return inflater.inflate(R.layout.choose_effect_fragment, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        recyclerView.setLayoutManager(new GridLayoutManager(getContext(), getResources().getInteger(R.integer.generator_type_adapter_grid_items)));
        recyclerView.setHasFixedSize(true);
        presenter.getEffectTypes();
    }

    @Override
    public void onSelected() {
        if (getView() != null) {
            presenter.getEffectTypes();
        }
    }

    @Override
    public void showTypes(@NonNull GeneratorType[] types, @Nullable GeneratorType selectedType, @NonNull GeneratorType targetType) {
        recyclerView.setAdapter(new GeneratorTypeAdapter(adapterImageLoader, types, selectedType, targetType, this, true));
    }

    @Override
    public void onSelected(@NonNull GeneratorType type) {
        logger.log(this, "onSelected: " + type.name());
        presenter.chooseEffectType(type);
    }

    @Override
    public void onDeselected() {
        logger.log(this, "onDeselected");
        presenter.chooseEffectType(null);
    }

    @Override
    public void onEditSelected() {

    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.onDetach();
    }
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/view/fragments/ChooseEffectFragment.kt
package com.stedi.randomimagegenerator.app.view.fragments

import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import butterknife.BindView
import com.stedi.randomimagegenerator.app.R
import com.stedi.randomimagegenerator.app.model.data.GeneratorType
import com.stedi.randomimagegenerator.app.presenter.interfaces.ChooseEffectPresenter
import com.stedi.randomimagegenerator.app.view.adapters.GeneratorTypeAdapter
import com.stedi.randomimagegenerator.app.view.components.BaseViewModel
import com.stedi.randomimagegenerator.app.view.fragments.base.GenerationFragment
import timber.log.Timber
import javax.inject.Inject

class ChooseEffectFragmentModel : BaseViewModel<ChooseEffectFragment>() {
    @Inject lateinit var presenter: ChooseEffectPresenter

    override fun onCreate(view: ChooseEffectFragment) {
        Timber.d("ChooseEffectFragmentModel onCreate")
        view.generationComponent.inject(this)
    }
}

class ChooseEffectFragment : GenerationFragment(), GeneratorTypeAdapter.ClickListener, ChooseEffectPresenter.UIImpl {

    private lateinit var viewModel: ChooseEffectFragmentModel

    @BindView(R.id.choose_effect_fragment_recycler_view) lateinit var recyclerView: RecyclerView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        viewModel = ViewModelProviders.of(this).get(ChooseEffectFragmentModel::class.java)
        viewModel.init(this)

        viewModel.presenter.onAttach(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        return inflater.inflate(R.layout.choose_effect_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        recyclerView.layoutManager = GridLayoutManager(context, resources.getInteger(R.integer.generator_type_adapter_grid_items))
        recyclerView.setHasFixedSize(true)
        viewModel.presenter.getEffectTypes()
    }

    override fun onSelected() {
        if (view != null) {
            viewModel.presenter.getEffectTypes()
        }
    }

    override fun showTypes(types: Array<GeneratorType>, selectedType: GeneratorType?, targetType: GeneratorType) {
        recyclerView.adapter = GeneratorTypeAdapter(activity!!, types, selectedType, targetType, this, true)
    }

    override fun onSelected(type: GeneratorType) {
        Timber.d("onSelected: ${type.name}")
        viewModel.presenter.chooseEffectType(type)
    }

    override fun onDeselected() {
        Timber.d("onDeselected")
        viewModel.presenter.chooseEffectType(null)
    }

    override fun onDestroy() {
        super.onDestroy()
        viewModel.presenter.onDetach()
    }

    override fun onEditSelected() {}
}